Stripe/STPRedirectContext.m

```diff
@@ -48,16 +68,64 @@ - (void)dealloc {
     [self unsubscribeFromNotificationsAndDismissPresentedViewControllers];
 }
 
-- (void)startRedirectFlowFromViewController:(UIViewController *)presentingViewController {
+- (void)performAppRedirectIfPossibleWithCompletion:(STPNativeRedirectCompletionBlock)onCompletion {
     FAUXPAS_IGNORED_IN_METHOD(APIAvailability)
-    if ([SFSafariViewController class] != nil) {
-        [self startSafariViewControllerRedirectFlowFromViewController:presentingViewController];
+
+    if (self.state == STPRedirectContextStateNotStarted) {
+        NSURL *nativeUrl = [self nativeRedirectURLForSource:self.source];
+        if (!nativeUrl) {
+            onCompletion(NO);
+            return;
+        }
+
+        // Optimistically start listening in case we get app switched away.
+        // If the app switch fails we'll undo this later
+        _state = STPRedirectContextStateInProgress;
+        [self subscribeToUrlAndForegroundNotifications];
+
+        UIApplication *application = [UIApplication sharedApplication];
+        if ([application respondsToSelector:@selector(openURL:options:completionHandler:)]) {
+
+            WEAK(self);
+            [application openURL:nativeUrl options:@{} completionHandler:^(BOOL success) {
+                if (!success) {
+                    STRONG(self);
+                    self->_state = STPRedirectContextStateNotStarted;
+                    [self unsubscribeFromNotifications];
+                }
+                onCompletion(success);
+            }];
+        }
+        else {
+            _state = STPRedirectContextStateInProgress;
+            BOOL opened = [application openURL:nativeUrl];
```

Comment:
OOC, do you know if this behaves differently from the new `openURL:options:completionHandler:` method? E.g. does the returned success value here always match the success value in the async completion block?

Reply:
As far as I know it's always the same, it's described the same in the docs. The main difference in the new API afaict is the options and the asyncronicity.